src/shell.js

```diff
@@ -34,7 +34,8 @@ var Module = typeof {{{ EXPORT_NAME }}} != 'undefined' ? {{{ EXPORT_NAME }}} : {
 #endif // USE_CLOSURE_COMPILER
 
 #if POLYFILL
-#if WASM_BIGINT && MIN_SAFARI_VERSION < 150000
+#if WASM_BIGINT && MIN_SAFARI_VERSION < 140100
+// TODO(https://github.com/emscripten-core/emscripten/issues/23184): Fix this back to 150000
```

Comment:
I don't follow these changes. This will make the polyfill be added in fewer cases - why is that correct?

Reply:
The point of the general messing with safari versions ( in feature_matrix itself) is to turn BIGINT on by default without affecting other features. This one here is just so that we won't default to having BIGINT on with this polyfill.

Reply:
Ok, thanks, I see. So the goal is to split out the safari version change? If so, I wonder if we can do that in this PR? The complexity of splitting it seems relatively high to me.

Reply:
Well, the real goal is to avoid changing all 3 settings in a single commit. But the commits also tie the settings more directly to the browser versions. So the simplest way to make enabling the features independent from each other (and independent from other consequences of raising the default version target) was to lower the version target for each feature separately (which should be a 2-line change for this feature, and 1-line for the others).

Reply:
Hmm, sorry, but why was that the real goal? Naively I'd think that adjusting the browser versions would be the natural way to enable features by default, so it fits in this PR, so I'm still missing something, sorry.

Reply:
I just worry about having this commit appear in bisection, say. This line here will include the polyfill using an incorrect rule, so it seems like it could be confusing, at such times. Thinking about it, how about this idea:

1. Add some fake Safari versions, so we have a linear progression. 14.7 adds bigint, 14.8 adds another, 14.9 adds yet another. No such real versions exist in reality, so users will never do `MIN_SAFARI=14.7`, making this unobservable.
2. We bump Safari to 14.7 in this PR, enabling only bigint. Later PRs bump to 14.8 etc.

This way we have a linear progression of features that is stable across commits (at least after the first commit that adds these fake versions).

Reply:
Yeah, the main thing was that raising the browser version would enable multiple features at once.
Having his appear in bisection is a feature, not a bug :). Also it has the effect that the PR that raises the default version also includes the relevant code and test changes. It seems like that would be helpful for debugging issues (as opposed to a bisection finding a change that just bumps a version number and then having to go to another PR to see the actual recent code change).
It seemed to me that the intended effect of the commit is clear, and each of the 2 oddities (incorrect versions, with links to a bug describing what is happening) seems clearer than introducing fake Safari versions. But maybe that's just me.

Another option (made possible by the fact that the features aren't really hooked up to browser versions correctly) would be to just bump the default version first (which wouldn't do much currently), and then land the PRs that hook each feature up to the browser version feature matrix. I'm not sure if that would be better, but I think it could work.

Reply:
If it really is the polyfill that worries you, Sam and I discussed in an earlier comment that we could just remove the polyfill and don't claim to support BigInt on Safari 14.1. The features page on https://webassembly.org/features/ actually claims this already because when I wrote it, I wasn't even aware of the polyfill. Or we could just leave this code as-is (with the effect that by default Emscripten will enable wasm-bigint and include the polyfill until we bump the default version).

Reply:
If it's just the polyfill then I don't feel too strongly. But are there not other changes in this PR that have the only goal of working around the normal feature detection/browser version scheme? My impression was that moving to 14.7, 14.8 etc. would simplify some of the features logic changes here, though perhaps those are needed for other reasons that I've missed.

Reply:
There is one other, in feature_matrix itself (which controls whether the feature is turned on or not (also with a link to #23184). This is of course not working around the versioning scheme, but temporarily modifying it. I don't think moving to fake versions would simplify the feature logic changes, since I don't think we'd want to keep the fake versions in the codebase; so either way we make a temporary change to the versioning logic, and then back it out soon (I'm still thinking we should do it before the next release).

Reply:
Since its temporary either way I don't feel strongly.

Reply:
Hm. I guess after the release we aren't really bisecting like you do with revisions, we're just testing flags to find out what breaks. So the options are testing by enabling or disabling the feature flags, or by setting the fake MIN_SAFARI version numbers and seeing what changes. But at that point the feature flags will still be there, so it's not clear to me what keeping the fake safari versions adds.

Reply:
> Hm. I guess after the release we aren't really bisecting like you do with revisions, we're just testing flags to find out what breaks.

Was this in reference to my mention of bisecting on `MIN_SAFARI`?

If so then my point there was that if we add Safari 14.7, 14.8 etc., avoiding a single version that adds multiple features, then bisecting `MIN_SAFARI` in the future becomes more useful. I agree that using the feature flags could also be done instead, but that requires one to know which flags are relevant and manually set them up, while bisecting on `MIN_SAFARI` can be done by just mindless bisection on a range of numbers. Like mindless bisection on commits, that sounds useful.

Anyhow this is a minor advantage of 14.7, 14.8 imo. The larger benefit as it I see it is that once we decide on the 14.7 etc. scheme, it is permanent: we have nothing to revert in a later PR, and we have no odd things that show up during bisection that were temporary before being undone. It is a design that "smooths" out safari versions to make them incremental. If browsers enable multiple features in a single release in the future then the same idea could be used again, that is, I don't see this as a one-off hack: it is a principled approach to this problem, as I see it.

But, again, if no one else sees value in this, that's fine :smile: I'm ok with the alternative, the difference is not huge here.

Reply:
Yeah, I think mainly where we disagree is that I don't really like the idea of keeping fake browser versions around in the codebase. Bisecting on browser versions isn't really mindless if you have to know that the fake version numbers exist and what they are for, and what they can be set to.

Reply:
Fair enough, sgtm. Maybe I like the idea of smoothing the history because it sounds like a math operation... but I agree fake versions have the downside you mention too.